assimilation_code/modules/assimilation/assim_tools_mod.pf.f90

```diff
@@ -1447,109 +1719,131 @@ subroutine filter_assim(ens_handle, obs_ens_handle, obs_seq, keys,           &
    OBS_UPDATE: do j = 1, num_close_obs
       obs_index = close_obs_ind(j)
 
-      if (my_obs_indx(obs_index) <= i) cycle OBS_UPDATE
```

Comment:
this line causes the loop to only process and update expected obs values which have not yet been assimilated.  it's missing from the updated code.  if the PF needs to loop over all obs every time then it needs an if (filter_kind /= 9) and this test to cycle for obs which have already been processed.  if it doesn't need to do anything with obs already assimilated, it's doing a lot more work than it needs to.

Reply:
this one is a science question.  i couldn't tell from the code if this loop had to cycle over all obs each time or not.  in the other filter kinds that's wasted work to update forward operator values for obs that have already been assimilated and won't be used again.  if this test was removed on purpose then a comment about adding the cycle code back in if the filter_kinds are integrated might be a good idea.

Reply:
Because the local PF now iterates multiple times over observations, it still needs to update obs-space priors for measurements that are already assimilated. We can, however, add this line back in for the last iteration.

Reply:
added a note about this line